libs/community/langchain_community/chat_models/vllm.py

```diff
@@ -0,0 +1,255 @@
+
+Here is the output schema you must conform to:
+```
+{schema}
+```"""  # noqa E501
+
+REGEX_FORMAT_INSTRUCTIONS = """The output should be formatted as a string that conforms to the regex below.
+
+As an example, for the regex `[a-z]+` the string `abc` is a well-formatted instance of the regex. The string `123` is not well-formatted.
+
+Here is the output regex you must conform to:
+```
+{schema}
+```"""  # noqa E501
+
+CHOICE_FORMAT_INSTRUCTIONS = """The output should be one of the following choices:
+```
+{schema}
+```"""
+
+GRAMMAR_FORMAT_INSTRUCTIONS = """The output should be a string that conforms to the grammar below.
+
+As an example, for the grammar
+```
+?start: expression
+
+?expression: term (("+" | "-") term)*
+
+?term: factor (("*" | "/") factor)*
+
+?factor: NUMBER
+        | "-" factor
+        | "(" expression ")"
+
+%import common.NUMBER
+```
+the string `2*3` is a well-formatted instance of the grammar. The string "two times three" is not well-formatted.
+
+Here is the output grammar you must conform to:
+```
+{schema}
+```"""  # noqa E501
+
+
+_BM = TypeVar("_BM", bound=BaseModel)
+_StructuredInput = Union[Dict[str, Any], str, List[str], Type[_BM]]
+_StructuredOutput = Union[Dict, str, _BM]
+
+
+class ChatVLLMOpenAI(ChatOpenAI):
+    """vLLM OpenAI-compatible API client"""
+
+    openai_api_key: Optional[SecretStr] = Field(default="dummy", alias="api_key")
+    """Automatically inferred from env var `OPENAI_API_KEY` if not provided."""
+    openai_api_base: Optional[str] = Field(
+        default="http://localhost:8000/v1", alias="base_url"
+    )
+    """Base URL path for API requests, leave blank if not using a proxy or service 
+        emulator."""
+    extra_body: Optional[Dict[str, Any]] = Field(default={})
+    """Extra body parameters to pass to the API request."""
+
+    @property
+    def _default_params(self) -> Dict[str, Any]:
+        """Get the default parameters for calling vllm."""
+        return {**super()._default_params, "extra_body": self.extra_body}
+
+    @property
+    def _llm_type(self) -> str:
+        """Return type of llm."""
+        return "vllm-openai"
+
+    @staticmethod
+    def _get_instructions(schema_str: str, guided_mode: str) -> str:
+        """Get the instructions to insert into the input."""
+        if guided_mode == "guided_json":
+            return "\n" + JSON_FORMAT_INSTRUCTIONS.format(schema=schema_str)
+        elif guided_mode == "guided_regex":
+            return "\n" + REGEX_FORMAT_INSTRUCTIONS.format(schema=schema_str)
+        elif guided_mode == "guided_choice":
+            return "\n" + CHOICE_FORMAT_INSTRUCTIONS.format(schema=schema_str)
+        elif guided_mode == "guided_grammar":
+            return "\n" + GRAMMAR_FORMAT_INSTRUCTIONS.format(schema=schema_str)
+        else:
+            raise ValueError(f"Unsupported guided_mode {guided_mode}")
+
+    @staticmethod
+    def _insert_instructions(
```

Comment:
I would appreciate some input on how to improve this function.
I tried to handle a couple of possible `LanguageModelInput`s, but I don't expect this to be exhaustive.
Is there a better way to append instructions to an incoming `LanguageModelInput`?

Reply:
Note that without this insertion of instructions, the `with_structured_output` function will not behave as expected. It really is crucial for the model to understand what output is expected. If the `insert_instructions` step would be removed, the output would still follow the provided schema (because vLLM will force the output to be compliant), but the quality of the generation would likely be very low.